sys/initrd.c

```diff
@@ -0,0 +1,193 @@
+#include <common.h>
+#include <malloc.h>
+#include <vm.h>
+#include <vm_map.h>
+#include <stdc.h>
+#include <initrd.h>
+#include <vnode.h>
+#include <mount.h>
+
+static MALLOC_DEFINE(mpool, "cpio mem_pool");
+
+static vm_addr_t rd_start;
+static vm_addr_t rd_size;
+
+
+vm_addr_t get_rd_start()
+{
+    return rd_start;
+}
+
+vm_addr_t get_rd_size()
+{
+    return rd_size;
+}
+
+int hex_atoi(char *s, int n)
+{
+    int res = 0;
+    for(int i = 0; i < n; i++)
+    {
+        int add = (s[i] <= '9') ? s[i]-'0' : s[i]-'A'+10;
+        res *= 16;
+        res += add;
+    }
+    return res;
+}
+
+int oct_atoi(const char *s, int n)
+{
+    int res = 0;
+    for(int i = 0; i < n; i++)
+    {
+        int add = s[i]-'0';
+        res *= 8;
+        res += add;
+    }
+    return res;
+}
+
+void ramdisk_init(vm_addr_t _rd_start, vm_addr_t _rd_size)
+{
+    rd_start = _rd_start;
+    rd_size = _rd_size;
+}
+
+int get_name_padding(int offset)
+{
+    return (4 - (offset % 4));
+}
+
+int get_file_padding(int offset)
+{
+    return (4 - (offset % 4)) % 4;
+}
+
+void dump_cpio_stat(cpio_file_stat_t *stat)
+{
+    kprintf("c_magic: %o\n", stat->c_magic);
+    kprintf("c_ino: %d\n", stat->c_ino);
+    kprintf("c_mode: %d\n", stat->c_mode);
+    kprintf("c_uid: %d\n", stat->c_uid);
+    kprintf("c_gid: %d\n", stat->c_gid);
+    kprintf("c_nlink: %zu\n", stat->c_nlink);
+    kprintf("c_mtime: %ld\n", stat->c_mtime);
+    kprintf("c_filesize: %ld\n", stat->c_filesize);
+    kprintf("c_dev_maj: %ld\n", stat->c_dev_maj);
+    kprintf("c_dev_min: %ld\n", stat->c_dev_min);
+    kprintf("c_rdev_maj: %ld\n", stat->c_rdev_maj);
+    kprintf("c_rdev_min: %ld\n", stat->c_rdev_min);
+    kprintf("c_namesize: %u\n", stat->c_namesize);
```

Comment:
Travis still uses the `mips-mti-elf` toolchain, which as a different definition of `size_t`. To avoid issues with printing `size_t`s, use the `%zu` format.